staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/interface.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/cel"
+
+	"github.com/google/cel-go/common/types/ref"
+)
+
+type FailurePolicy string
+
+const (
+	Fail   FailurePolicy = "Fail"
+	Ignore FailurePolicy = "Ignore"
+)
+
+type EvaluatorFunc func(a admission.Attributes, params *unstructured.Unstructured) []PolicyDecision
+
+// Dependency Injected into the PolicyDefinition's `Compile` function to assist
+// with converting types and values to/from CEL-typed values
+type ObjectConverter interface {
+	// Given a GVR, look up the openapi or JSONSchemaProps, structural schema, etc.
+	// and compile it into something that can be used to turn objects into CEL
+	// values
+	DeclForResource(gvr schema.GroupVersionResource) *cel.DeclType
+
+	// Object may be a typed native object or an unstructured object
+	ValueForObject(value runtime.Object, decl *cel.DeclType) *ref.Val
+}
+
+// An interface for internal policy binding type. Implemented by mock/testing types,
+// and to be implemented by the public API types once they have completed API review.
+//
+// The interface closely mirrors the format and functionality of the
+// PolicyDefinition proposed in the KEP.
+type PolicyDefinition interface {
+	runtime.Object
+
+	// returns namespace/name
+	GetNamespace() string
+	GetName() string
+
+	// says whether this policy definition matches the provided admission resource request
+	Matches(a admission.Attributes) bool
+
+	Compile(
+		// Definition is provided with a converter which may be used by the
+		// return evaluator function to convert objects into CEL-typed objects
+		objectConverter ObjectConverter,
+	) (EvaluatorFunc, error)
```

Comment:
Will it be good to have structured err here? We are going to have a status err attached to each Validation. I am not sure how we are going to leverage it if the return is error.

Reply:
Since we aren't 100% sure about error reporting implementation details I'd vote for leaving this as `error`. We will be able to change later if it does not suit us well. We can also an optional cast if necessary i.e. (`if structured, ok := err.(StatusError); ok`).